cpp/src/parquet/arrow/reader.cc

```diff
@@ -453,229 +483,226 @@ class NestedListReader : public ColumnReaderImpl {
-        return Status::NotImplemented(
-            "Nested data conversions not implemented for chunked array outputs");
-    }
+  Status LoadBatch(int64_t number_of_records) final {
+    return item_reader_->LoadBatch(number_of_records);
+  }
 
+  Status BuildArray(int64_t number_of_slots,
+                    std::shared_ptr<ChunkedArray>* out) override {
     const int16_t* def_levels;
     const int16_t* rep_levels;
     int64_t num_levels;
     RETURN_NOT_OK(item_reader_->GetDefLevels(&def_levels, &num_levels));
     RETURN_NOT_OK(item_reader_->GetRepLevels(&rep_levels, &num_levels));
-    std::shared_ptr<Array> result;
-    RETURN_NOT_OK(ReconstructNestedList(item_chunk, field_, max_definition_level_,
-                                        max_repetition_level_, def_levels, rep_levels,
-                                        num_levels, ctx_->pool, &result));
+    std::shared_ptr<Buffer> validity_buffer;
+    ::parquet::internal::ValidityBitmapInputOutput validity_io;
+    validity_io.values_read_upper_bound = number_of_slots;
+    if (field_->nullable()) {
+      ARROW_ASSIGN_OR_RAISE(validity_buffer, AllocateBitmap(number_of_slots, ctx_->pool));
+
+      validity_io.valid_bits = validity_buffer->mutable_data();
+    }
+    ARROW_ASSIGN_OR_RAISE(
+        std::shared_ptr<Buffer> lengths_buffer,
+        AllocateBuffer(sizeof(IndexType) * std::max(int64_t{2}, number_of_slots + 1),
+                       ctx_->pool));
+    // ensure zero initialization in case we have reached a zero length list (and
+    // because first entry is always zero).
+    IndexType* length_data = reinterpret_cast<IndexType*>(lengths_buffer->mutable_data());
+    std::fill(length_data, length_data + 2, 0);
+    BEGIN_PARQUET_CATCH_EXCEPTIONS
+    ::parquet::internal::DefRepLevelsToList(def_levels, rep_levels, num_levels,
+                                            level_info_, &validity_io, length_data);
+    END_PARQUET_CATCH_EXCEPTIONS
+    // We might return less then the requested slot (i.e. reaching an end of a file)
+    // ensure we've set all the bits here.
+    if (validity_io.values_read < number_of_slots) {
+      // + 1  because arrays lengths are values + 1
+      std::fill(length_data + validity_io.values_read + 1,
+                length_data + number_of_slots + 1, 0);
+      if (validity_io.valid_bits != nullptr &&
+          BitUtil::BytesForBits(validity_io.values_read) <
+              BitUtil::BytesForBits(number_of_slots)) {
+        std::fill(validity_io.valid_bits + BitUtil::BytesForBits(validity_io.values_read),
+                  validity_io.valid_bits + BitUtil::BytesForBits(number_of_slots), 0);
+      }
+    }
+
+    RETURN_NOT_OK(item_reader_->BuildArray(length_data[validity_io.values_read], out));
+    ARROW_ASSIGN_OR_RAISE(std::shared_ptr<ArrayData> item_chunk, ChunksToSingle(**out));
+
+    std::vector<std::shared_ptr<Buffer>> buffers{
+        validity_io.null_count > 0 ? validity_buffer : std::shared_ptr<Buffer>(),
```

Comment:
Can just write `nullptr`, I think.

Reply:
yes, I thought this was causing problems with type inference at some pont.